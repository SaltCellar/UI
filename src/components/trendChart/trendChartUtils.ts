import { Report } from 'api/reports';
import format from 'date-fns/format';
import getDate from 'date-fns/get_date';
import startOfMonth from 'date-fns/start_of_month';
import { getComputedReportItems } from 'utils/getComputedReportItems';
import { SortDirection } from 'utils/sort';

export interface TrendChartDatum {
  x: number;
  y: number;
  date: string | number;
}

export type DatumValueFormatter = (value: number) => string | number;

export function transformReport(report: Report): TrendChartDatum[] {
  if (!report) {
    return [];
  }

  return getComputedReportItems({
    report,
    idKey: 'date',
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  }).reduce<TrendChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    const nextItem: TrendChartDatum = {
      x: getDate(d.id),
      y: prevValue + d.total,
      date: d.id,
    };
    return [...acc, nextItem];
  }, []);
}

export function getDatumDateRange(datums: TrendChartDatum[]): [Date, Date] {
  if (!datums.length) {
    const today = new Date();
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].date + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].date + 'T00:00:00');
  return [start, end];
}

export function getDateRangeString(datums: TrendChartDatum[]) {
  const [start, end] = getDatumDateRange(datums);
  const monthName = format(start, 'MMM');
  const startDate = getDate(start);
  const endDate = getDate(end);
  return `${monthName} ${getDate(start)}${
    startDate !== endDate ? ` - ${endDate}` : ''
  }`;
}

export function getTooltipLabel(
  datum: TrendChartDatum,
  formatValue: DatumValueFormatter
) {
  if (!datum.date) {
    return '';
  }
  const date = format(datum.date, 'MMM D YYYY');
  return `${date}: ${formatValue(datum.y)}`;
}
