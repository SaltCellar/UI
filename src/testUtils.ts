import { ReactWrapper, ShallowWrapper } from 'enzyme';
import { testIdProp } from 'testIds';

export function wait() {
  return new Promise(resolve => setImmediate(resolve));
}

export function findByTestId(
  view: ShallowWrapper | ReactWrapper,
  testId: string
) {
  return (view as ShallowWrapper)
    .findWhere(node => node.prop(testIdProp) === testId)
    .first();
}
export function mockDate() {
  const constantDate = new Date(2018, 0, 1, 0);
  (global as any).Date = class extends Date {
    constructor(dateString: string) {
      super(dateString);
      if (!dateString) {
        return constantDate;
      }
    }
  };
}
