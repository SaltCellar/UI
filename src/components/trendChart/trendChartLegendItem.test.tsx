import { shallow } from 'enzyme';
import React from 'react';

import { mockDate } from 'testUtils';
import {
  TrendChartLegendItem,
  TrendChartLegendItemProps,
} from './trendChartLegendItem';
import * as utils from './trendChartUtils';

mockDate();

const props: TrendChartLegendItemProps = {
  data: [createDatum('2018-01-10'), createDatum('2018-01-15')],
};

test('range is formated with start and end date', () => {
  const view = shallow(<TrendChartLegendItem {...props} />);
  expect(view.text()).toMatchSnapshot();
});

test('range is formated with start and end date for an empty report', () => {
  const view = shallow(<TrendChartLegendItem data={[]} />);
  expect(view.text()).toMatchSnapshot();
});

function createDatum(date: string): utils.TrendChartDatum {
  return {
    date,
    x: 1,
    y: 1,
    units: 'units',
  };
}
