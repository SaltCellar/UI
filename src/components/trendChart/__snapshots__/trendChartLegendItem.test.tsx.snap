// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`range is formated with start and end date 1`] = `"Jan 10 - 15"`;

exports[`range is formated with start and end date for an empty report 1`] = `""`;
